[SPARK-25440][SQL] Dumping query execution info to a file

MaxGekk opened the pull request:
## What changes were proposed in this pull request?

In the PR, I propose new method for debugging queries by dumping info about their execution to a file. It saves logical, optimized and physical plan similar to the `explain()` method + generated code. One of the advantages of the method over `explain` is it doesn't truncate output and doesn't not materializes full output as one string in memory which can cause OOMs.

## How was this patch tested?

Added a test which checks that new method dumps correct info about a query.

MaxGekk committed:
Stub implementation and a test

MaxGekk committed:
Saving all plans to file

MaxGekk committed:
Output attributes

MaxGekk committed:
Output whole stage codegen

MaxGekk committed:
Reusing codegenToOutputStream

MaxGekk committed:
Code de-duplication

MaxGekk committed:
Do not truncate fields

SparkQA commented:
**[Test build #96093 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96093/testReport)** for PR 22429 at commit [`2ee75bc`](https://github.com/apache/spark/commit/2ee75bcd4d495eb6031581ad7a38e757c254175b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96093/
Test FAILed.

MaxGekk committed:
Moving the test up because previous one leaved a garbage

SparkQA commented:
**[Test build #96097 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96097/testReport)** for PR 22429 at commit [`9b2a3e6`](https://github.com/apache/spark/commit/9b2a3e664564182e325ef08785c998c6ff9b5367).

MaxGekk commented:
@rednaxelafx Please, take a look at the PR.

SparkQA commented:
**[Test build #96097 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96097/testReport)** for PR 22429 at commit [`9b2a3e6`](https://github.com/apache/spark/commit/9b2a3e664564182e325ef08785c998c6ff9b5367).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96097/
Test PASSed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:
```diff
@@ -469,7 +470,13 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
+    bos.toString
+  }
+
+  def treeString(os: OutputStream, verbose: Boolean, addSuffix: Boolean): Unit = {
```
Can you please use a `java.io.Writer` or something else you can directly write a string to? You are now using `getBytes()` everywhere and that is far from cheap because it needs to encode the chars and allocate a byte array for each string.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:
```diff
@@ -250,5 +253,35 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
+     */
+    def toFile(path: String): Unit = {
+      val filePath = new Path(path)
+      val fs = FileSystem.get(filePath.toUri, sparkSession.sparkContext.hadoopConfiguration)
```
Why use the hadoop configuration of the `SparkContext`? It is probably better to use the one that `sparkSession.sessionState.newHadoopConf()` provides.

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/QueryExecutionSuite.scala:
```diff
@@ -16,11 +16,33 @@
 class QueryExecutionSuite extends SharedSQLContext {
+  test("dumping query execution info to a file") {
+    withTempDir { dir =>
+      val path = dir.getCanonicalPath + s"/plans.txt"
```
We don't need string interpolation here.

MaxGekk committed:
Removing string interpolation in the test

MaxGekk committed:
Getting Hadoop's conf from session state

MaxGekk committed:
Using java.io.Writer

MaxGekk committed:
Using java.io.Writer

SparkQA commented:
**[Test build #96109 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96109/testReport)** for PR 22429 at commit [`ce2c086`](https://github.com/apache/spark/commit/ce2c08688bb8b51e97f686c95279a5f42b52116a).

SparkQA commented:
**[Test build #96109 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96109/testReport)** for PR 22429 at commit [`ce2c086`](https://github.com/apache/spark/commit/ce2c08688bb8b51e97f686c95279a5f42b52116a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96109/
Test PASSed.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:
```diff
@@ -250,5 +254,36 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
+      val writer = new BufferedWriter(new OutputStreamWriter(fs.create(filePath)))
+
+      try {
+        SparkEnv.get.conf.set(Utils.MAX_TO_STRING_FIELDS, Int.MaxValue.toString)
```
It is generally a bad idea to change this conf as people expect that it is immutable. Also this change has some far reaching consequences, others will now also be exposed to a different `Utils.MAX_TO_STRING_FIELDS` value when calling `explain()`. Can you please just pass the parameter down the tree?

MaxGekk replied:
Herman, I am trying to do that but I am already doubt it is right ways. I have to add new argument to `verboseString` and `simpleString` which is broadly used everywhere.

MaxGekk replied:
> It is generally a bad idea to change this conf as people expect that it is immutable.

Maybe it is generally a bad idea but the purpose of the `toFile` is to call it during debug of an issue. Currently when someone need to output all fields, he/she has to set `spark.debug.maxToStringFields` before `explain()` and revert it back after `explain()`. If you want, I can add a comment which tells to an user that the `toFile` changes the config during its invoke.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:
```diff
@@ -250,5 +254,36 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
+
+      try {
+        SparkEnv.get.conf.set(Utils.MAX_TO_STRING_FIELDS, Int.MaxValue.toString)
+        writer.write("== Parsed Logical Plan ==\n")
```
Can we combine this entire block with what is done in the `toString()` method?

MaxGekk replied:
Combined

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:
```diff
@@ -469,7 +470,17 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
 
   def treeString(verbose: Boolean, addSuffix: Boolean = false): String = {
-    generateTreeString(0, Nil, new StringBuilder, verbose = verbose, addSuffix = addSuffix).toString
+    val baos = new ByteArrayOutputStream()
```
In this particular method, there is no benefits. This was changed to reused the method which accepts `OutputStream` instead of `StringBuilder`. Benefit of `OutputStream` over `StringBuilder` is no full materialization in memory and no string size limit.

MaxGekk replied:
This is done

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into plan-to-file

MaxGekk committed:
Using StringWriter

MaxGekk committed:
Removing unneeded buffering and flushing

MaxGekk committed:
Code de-duplication among toString and toFile

rednaxelafx commented on sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:
```diff
@@ -189,23 +192,32 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
+    writeOrError(writer)(logical.treeString(_, verbose = true, addSuffix = false))
+    writer.write("== Analyzed Logical Plan ==\n")
+    writeOrError(writer) { writer =>
+      analyzed.output.foreach(o => writer.write(s"${o.name}: ${o.dataType.simpleString}"))
```
If you want to get the best out of this approach, it might be better to avoid string interpolation here and do the explicit `writer.write` calls for the things you're interpolating on.

rednaxelafx reviewed (commented):
Thank you so much for the proposal, Max!

While I really really like the feature you're proposing (yes please get them in!), as they're really useful for debugging / diagnosis, there are a few things in the current implementation that might be deal breakers for me:

1. Similar to what @hvanhovell mentioned, if we want to have a clean feature at the end, it would be better to pass the max fields argument down the call chain instead of changing the Spark conf.
Setting a Spark conf is a process-wide global change that will affect everybody, including code running in other threads. A save-use-restore sequence would only work for a single threaded scenario, and would otherwise be problematic in a multi-threaded scenario.

It would definitely be a lot more code change to pass that argument down, but I think it's worth the change (because the existing code before this change is too rigid to begin with anyway).

2. The use of `java.io.StringWriter`. I totally understand the motivation to use `StringWriter` for the regular path so that both the regular and file-dumping paths can share the same implementation code via the `Writer` interface.

But if you take a look at how `StringWriter` is implemented in OpenJDK8u: http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/2660b127b407/src/share/classes/java/io/StringWriter.java#l43
```java
public class StringWriter extends Writer {

    private StringBuffer buf;
```
It's actually backed by a `StringBuffer`, which is worse than a `StringBuilder` in terms of extra synchronization overhead implied.
(Yes, modern JVMs to try to optimize `StringBuffer`s via various techniques, including biased locking, lock elimination etc, but it's just less guaranteed to have good performance)
The best-performing way to do this is probably to implement a custom `Writer` that uses some sort of a "rope" implementation as the backing store, so that premature string concatenation is kept to a minimum. Whether or not we wanna go that way can be subject of further discussion, though.

SparkQA commented:
**[Test build #96154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96154/testReport)** for PR 22429 at commit [`71ff7d1`](https://github.com/apache/spark/commit/71ff7d1387fbe7d30299fe38471bce26fe73dad5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96154/
Test FAILed.

gatorsmile commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -99,10 +99,11 @@ private[spark] object Utils extends Logging {
    * by setting the 'spark.debug.maxToStringFields' conf in SparkEnv.
    */
   val DEFAULT_MAX_TO_STRING_FIELDS = 25
+  val MAX_TO_STRING_FIELDS = "spark.debug.maxToStringFields"
```
Can we move it to org.apache.spark.internal.config? I think our Core module should do the same thing like what we are doing in SQLConf.

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/execution/QueryExecutionSuite.scala:
```diff
@@ -16,11 +16,33 @@
 class QueryExecutionSuite extends SharedSQLContext {
+  test("dumping query execution info to a file") {
+    withTempDir { dir =>
+      val path = dir.getCanonicalPath + "/plans.txt"
```
Also add a few cases. For example, the path already exists? the path is not legal? and so on.

MaxGekk replied:
> Also add a few cases.

@gatorsmile I added a few tests. Please, take a look at them.

MaxGekk commented:
@rednaxelafx Thank you for such detailed comment.

> It would be better to pass the max fields argument down the call chain instead of changing the Spark conf.

Sure, it would be better. I just afraid I will break something by changing signature of methods like `simpleString`

> The best-performing way to do this is probably to implement a custom Writer

I am going to replace `StringWriter` by `StringBuilderWriter`. It should be faster though I am not sure it matters here.

MaxGekk committed:
Using StringBuilderWriter and fix tests

hvanhovell commented:
@MaxGekk please just modify simpleString it is internal API for this reason.

@rednaxelafx rope approach has the benefit that it does not create a ton of intermediate buffers. We could do that as well.

MaxGekk committed:
Do not change maxFields so far

MaxGekk committed:
Added tests

SparkQA commented:
**[Test build #96175 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96175/testReport)** for PR 22429 at commit [`d3fede1`](https://github.com/apache/spark/commit/d3fede192123bd490b5a6a875559fa43d43bfc3e).

MaxGekk committed:
Using StringBuilderWriter in treeString

MaxGekk committed:
Propagating numFields to truncatedString

MaxGekk committed:
Bug fix + test

SparkQA commented:
**[Test build #96178 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96178/testReport)** for PR 22429 at commit [`3324927`](https://github.com/apache/spark/commit/33249274cd9b2fa8ff5e934fe572123e562bf105).

MaxGekk committed:
Bug fix: passing maxFields to simpleString

SparkQA commented:
**[Test build #96175 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96175/testReport)** for PR 22429 at commit [`d3fede1`](https://github.com/apache/spark/commit/d3fede192123bd490b5a6a875559fa43d43bfc3e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96175/
Test FAILed.

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into plan-to-file

SparkQA commented:
**[Test build #96178 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96178/testReport)** for PR 22429 at commit [`3324927`](https://github.com/apache/spark/commit/33249274cd9b2fa8ff5e934fe572123e562bf105).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96178/
Test FAILed.

MaxGekk commented:
Tests failed on `PythonForeachWriterSuite`. Waiting for https://github.com/apache/spark/pull/22452

SparkQA commented:
**[Test build #96192 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96192/testReport)** for PR 22429 at commit [`24dbbba`](https://github.com/apache/spark/commit/24dbbbadc100bcc1b66ad1f43afe65d86401d835).

SparkQA commented:
**[Test build #96192 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96192/testReport)** for PR 22429 at commit [`24dbbba`](https://github.com/apache/spark/commit/24dbbbadc100bcc1b66ad1f43afe65d86401d835).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96192/
Test FAILed.

SparkQA commented:
**[Test build #96214 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96214/testReport)** for PR 22429 at commit [`24dbbba`](https://github.com/apache/spark/commit/24dbbbadc100bcc1b66ad1f43afe65d86401d835).

MaxGekk commented:
@hvanhovell May I ask you to look at this one more time.

MaxGekk commented:
@gatorsmile @rednaxelafx @HyukjinKwon @viirya @hvanhovell Could you review the PR, please.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala:
```diff
@@ -668,11 +670,19 @@ case class WholeStageCodegenExec(child: SparkPlan)(val codegenStageId: Int)
+      writer,
+      verbose,
+      s"*($codegenStageId) ",
+      false,
```
named boolean : `addSuffix = false`

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:
```diff
@@ -250,5 +265,22 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
+     */
+    def toFile(path: String): Unit = {
+      val filePath = new Path(path)
+      val fs = FileSystem.get(filePath.toUri, sparkSession.sessionState.newHadoopConf())
```
val fs = filePath.getFileSystem(spark.sessionState.newHadoopConf())

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:
```diff
@@ -250,5 +265,22 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
+    def toFile(path: String): Unit = {
+      val filePath = new Path(path)
+      val fs = FileSystem.get(filePath.toUri, sparkSession.sessionState.newHadoopConf())
+      val writer = new OutputStreamWriter(fs.create(filePath))
```
cc @zsxwing Could you help review this function?

zsxwing replied:
`val writer = new BufferedWriter(new OutputStreamWriter(fs.create(filePath), UTF_8))`

zsxwing commented on sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala:
```diff
@@ -196,7 +196,7 @@ case class RDDScanExec(
 
-  override def simpleString: String = {
+  override def simpleString(maxFields: Option[Int]): String = {
     s"$nodeName${Utils.truncatedString(output, "[", ",", "]")}"
```
`maxFields` should be used here.

zsxwing reviewed (changes requested):
`org.apache.spark.util.Utils.truncatedString` seems in a wrong project. Spark core doesn't have any schema or fields and I cannot see any chance that this method will be used in Spark core. Could you move it to the catalyst project. In addition, it would be great that you can change it to a SQL conf so that we can modify this config dynamically.

I also noticed that you forget to pass `maxFields` in several places. Can we change `maxFields: Option[Int]` to `maxFields: Int` so that the caller will never forget to pass it in and it should either get it from some place or use the default value.

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into plan-to-file

MaxGekk commented:
>  Can we change `maxFields: Option[Int]` to `maxFields: Int` so that the caller will never forget to pass it in and it should either get it from some place or use the default value.

@zsxwing Sure, but only replacing `Option[Int]` by `Int` cannot guarantee that. Need to remove default value to force a caller to pass a value to `simpleString()`

MaxGekk committed:
Passing parameters by names

MaxGekk committed:
Getting file system from file path

zsxwing commented:
@MaxGekk Make sense. Could you also try to remove the default value?

MaxGekk committed:
Using the buffered writer

MaxGekk committed:
Removing default value for maxFields in simpleString

MaxGekk committed:
Removing unnecessary signature of truncatedString

SparkQA commented:
**[Test build #96527 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96527/testReport)** for PR 22429 at commit [`4ec5732`](https://github.com/apache/spark/commit/4ec57326612adb2d650e425d03eae70366480242).

SparkQA commented:
**[Test build #96527 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96527/testReport)** for PR 22429 at commit [`4ec5732`](https://github.com/apache/spark/commit/4ec57326612adb2d650e425d03eae70366480242).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96527/
Test PASSed.

MaxGekk committed:
Minor improvement - passing maxFields by name

MaxGekk committed:
Moving truncatedString out of core

SparkQA commented:
**[Test build #96551 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96551/testReport)** for PR 22429 at commit [`90ff7b5`](https://github.com/apache/spark/commit/90ff7b543967d8f98eaa35dd456ef43d3866097f).

MaxGekk commented:
> Could you also try to remove the default value?

I have removed the default value `None` but still keep type as `Option[Int]`. @zsxwing Is it ok?

> org.apache.spark.util.Utils.truncatedString seems in a wrong project. ... Could you move it to the catalyst project. 

Moved to `catalyst`.

> In addition, it would be great that you can change it to a SQL conf so that we can modify this config dynamically.

Do you mean moving the `spark.debug.maxToStringFields` to `SQLConf` or adding another SQL config like `spark.sql.maxFieldsInTruncatedString`?

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:
```diff
@@ -167,6 +170,56 @@ package object util {
+
+  private[spark] def maxNumToStringFields = {
+    if (SparkEnv.get != null) {
+      SparkEnv.get.conf.getInt("spark.debug.maxToStringFields", DEFAULT_MAX_TO_STRING_FIELDS)
```
Since it is moved to sql, shall we name it as `spark.sql.debug.maxToStringFields`?

MaxGekk replied:
Renaming it can potentially break user's apps. Can we do that in current minor version?

HyukjinKwon replied:
I think it makes sense to me.

MaxGekk replied:
I have added new SQL config and I am testing it for now. My concern is how to combine old config, new config and passed parameter. I am going to take maximum of them like:
```scala
 private[spark] def maxNumToStringFields = {
    val legacyLimit = if (SparkEnv.get != null) {
      SparkEnv.get.conf.getInt("spark.debug.maxToStringFields", DEFAULT_MAX_TO_STRING_FIELDS)
    } else {
      DEFAULT_MAX_TO_STRING_FIELDS
    }
    val sqlConfLimit = SQLConf.get.maxToStringFields

    Math.max(sqlConfLimit, legacyLimit)
  }
```

SparkQA commented:
**[Test build #96551 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96551/testReport)** for PR 22429 at commit [`90ff7b5`](https://github.com/apache/spark/commit/90ff7b543967d8f98eaa35dd456ef43d3866097f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96551/
Test PASSed.

MaxGekk commented:
As far as I know, @hvanhovell, you would prefer replacing `Option[Int]` by just `Int` in `simpleString` and other similar places. I can do that but I have to read the config parameter in all those places where I pass `None` at the moment. I have counted ~ 26 places. @zsxwing Do you also prefer this approach?

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into plan-to-file

MaxGekk committed:
Adding SQL config to control maximum number of fields

MaxGekk committed:
Adding Spark Core config to control maximum number of fields

SparkQA commented:
**[Test build #96713 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96713/testReport)** for PR 22429 at commit [`2bf11fc`](https://github.com/apache/spark/commit/2bf11fcb1c22d7118c2bcb24cc279494ea953482).

SparkQA commented:
**[Test build #96713 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96713/testReport)** for PR 22429 at commit [`2bf11fc`](https://github.com/apache/spark/commit/2bf11fcb1c22d7118c2bcb24cc279494ea953482).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96713/
Test FAILed.

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into plan-to-file

MaxGekk committed:
Revert indentations

SparkQA commented:
**[Test build #96718 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96718/testReport)** for PR 22429 at commit [`bd331c5`](https://github.com/apache/spark/commit/bd331c5f8b2e3a0642cc9f818f400af9cbcf37ef).

SparkQA commented:
**[Test build #96718 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96718/testReport)** for PR 22429 at commit [`bd331c5`](https://github.com/apache/spark/commit/bd331c5f8b2e3a0642cc9f818f400af9cbcf37ef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96718/
Test PASSed.

MaxGekk commented:
@gatorsmile @zsxwing @hvanhovell @viirya @rednaxelafx @HyukjinKwon Please, take a look at the PR.

MaxGekk commented:
Is there anything which blocks the PR for now?

MaxGekk commented:
@zsxwing Please, have a look at the PR one more time.

boy-uber commented:
Great to see the code here to capture more information from the plan and save to file. We have similar needs with our Spark applications as well.

Adding my two cents:

1. Anyone feel the need to log the spark plan as a json or other structured representation? Current text based logging for spark plan is good for human, but hard to be processed by code/tool.

2. Could we generalize the solution a little bit? for example, writing the logical/physical plan json representation (or similar tree structure) to Spark listener, so people could write their own plugin to either write to file, or write to database, or write to kafka, etc.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:
```diff
@@ -455,21 +457,37 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
 
   /** ONE line description of this node. */
-  def simpleString: String = s"$nodeName $argString".trim
+  def simpleString(maxFields: Option[Int]): String = {
```
Please document the `maxFields` parameter. I am especially interested in what `None` represents here.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:
```diff
@@ -167,6 +172,58 @@ package object util {
+   * `spark.sql.debug.maxToStringFields`.
+   */
+  private[spark] def maxNumToStringFields: Int = {
+    val legacyLimit = if (SparkEnv.get != null) {
```
Just for context why do you want to retain the legacy behavior? It is probably fine to break it.

MaxGekk replied:
Taking into account that old config wasn't well documented and could use mostly in debugging, I think we can remove it in Spark 3.0. Initially the PR targeted to Spark 2.4, in the minor version removing a public config can break user apps potentially. If you are ok to remove it, I will do that.

MaxGekk replied:
I removed old core config and leaved only SQL config

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:
```diff
@@ -167,6 +172,58 @@ package object util {
+      sep: String,
+      end: String,
+      maxFields: Option[Int]): String = {
+    val maxNumFields = maxFields.getOrElse(maxNumToStringFields)
```
You should document this behavior.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:
```diff
@@ -189,23 +192,34 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
   }
 
+  private def writeOrError(writer: Writer)(f: Writer => Unit): Unit =
+    try f(writer) catch { case e: AnalysisException => writer.write(e.toString) }
```
Please use multiple lines here.

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into plan-to-file

MaxGekk commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -633,4 +633,14 @@ package object config {
+  private[spark] val MAX_TO_STRING_FIELDS =
+    ConfigBuilder("spark.debug.maxToStringFields")
+      .internal()
+      .doc("Maximum number of fields of sequence-like entries that can be converted to strings " +
```
I don't know how else I can describe all kind of classes where the parameter is applicable. If you have better words, you are welcome.

MaxGekk replied:
I am going to change it to `Maximum number of fields of a tree node that can be ...` for the SQL config.

MaxGekk committed:
Making writeOrError multi-line

MaxGekk committed:
Removing core config: spark.debug.maxToStringFields

MaxGekk committed:
Improving description of spark.sql.debug.maxToStringFields

MaxGekk commented on external/avro/src/main/scala/org/apache/spark/sql/avro/CatalystDataToAvro.scala:
```diff
@@ -52,7 +52,7 @@ case class CatalystDataToAvro(child: Expression) extends UnaryExpression {
 
-  override def simpleString: String = {
+  override def simpleString(maxFields: Option[Int]): String = {
     s"to_avro(${child.sql}, ${child.dataType.simpleString})"
```
passed

MaxGekk committed:
Limit number of fields in structs too

MaxGekk committed:
Description of simpleString of TreeNode.

MaxGekk committed:
Added description of maxFields param of truncatedString

MaxGekk committed:
Fix typo

MaxGekk committed:
Passing maxField

SparkQA commented:
**[Test build #97253 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97253/testReport)** for PR 22429 at commit [`e4567cb`](https://github.com/apache/spark/commit/e4567cbc01d58a17a1b4ba8a618463b8872a69a2).

MaxGekk committed:
Fix for the warning

SparkQA commented:
**[Test build #97256 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97256/testReport)** for PR 22429 at commit [`9b72104`](https://github.com/apache/spark/commit/9b721046c103f55d0abd4f65fd3e2e4b60be5ace).

SparkQA commented:
**[Test build #97256 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97256/testReport)** for PR 22429 at commit [`9b72104`](https://github.com/apache/spark/commit/9b721046c103f55d0abd4f65fd3e2e4b60be5ace).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97256/
Test FAILed.

SparkQA commented:
**[Test build #97253 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97253/testReport)** for PR 22429 at commit [`e4567cb`](https://github.com/apache/spark/commit/e4567cbc01d58a17a1b4ba8a618463b8872a69a2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97253/
Test PASSed.

boy-uber commented:
> @boy-uber the thing you are suggesting is a pretty big undertaking and beyond the scope of this PR.
> 
> If you are going to add structured plans to the explain output, you probably also want some guarantees about stability over multiple spark versions and you probably also want to be able to reconstruct the plan. Neither is easy. If you want to have this in Spark, then I suggest sending a proposal to the dev list.

Yeah, that is a larger change and may need more discussion. Your points about adding structured plans like that are great! Let me send a email to the dev list then! Thanks for the suggestion :)

MaxGekk committed:
Merge branch 'master' into plan-to-file

SparkQA commented:
**[Test build #97302 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97302/testReport)** for PR 22429 at commit [`9f1d11d`](https://github.com/apache/spark/commit/9f1d11df99ce37959229b2830b89e2a943d638f0).

MaxGekk commented:
@hvanhovell @zsxwing Could you look at this during the next a few days otherwise I will be able to come back to the PR in 3 weeks, please.

SparkQA commented:
**[Test build #97302 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97302/testReport)** for PR 22429 at commit [`9f1d11d`](https://github.com/apache/spark/commit/9f1d11df99ce37959229b2830b89e2a943d638f0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97302/
Test PASSed.

SparkQA commented:
**[Test build #97580 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97580/testReport)** for PR 22429 at commit [`9f1d11d`](https://github.com/apache/spark/commit/9f1d11df99ce37959229b2830b89e2a943d638f0).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97580/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #97600 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97600/testReport)** for PR 22429 at commit [`9f1d11d`](https://github.com/apache/spark/commit/9f1d11df99ce37959229b2830b89e2a943d638f0).

SparkQA commented:
**[Test build #97600 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97600/testReport)** for PR 22429 at commit [`9f1d11d`](https://github.com/apache/spark/commit/9f1d11df99ce37959229b2830b89e2a943d638f0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97600/
Test PASSed.

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into plan-to-file

# Conflicts:
#	external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDataToCatalyst.scala
#	external/avro/src/main/scala/org/apache/spark/sql/avro/CatalystDataToAvro.scala

SparkQA commented:
**[Test build #98324 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98324/testReport)** for PR 22429 at commit [`76f4248`](https://github.com/apache/spark/commit/76f424830418129c12a2a08d81f19377490c95eb).

MaxGekk commented:
> @MaxGekk I sent email to spark dev list about structured plan logging, but did not get any response. 

@boy-uber I guess It is better to speak about the feature to @bogdanrdc @hvanhovell @larturus

MaxGekk commented:
May I ask you @hvanhovell @zsxwing to review the PR one more time.

boy-uber commented:
> > @MaxGekk I sent email to spark dev list about structured plan logging, but did not get any response.
> 
> @boy-uber I guess It is better to speak about the feature to @bogdanrdc @hvanhovell @larturus

Thanks @MaxGekk for the contact list! I will ping them to gather more thoughts.

MaxGekk commented:
jenkins, retest this, please

SparkQA commented:
**[Test build #98461 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98461/testReport)** for PR 22429 at commit [`76f4248`](https://github.com/apache/spark/commit/76f424830418129c12a2a08d81f19377490c95eb).

SparkQA commented:
**[Test build #98461 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98461/testReport)** for PR 22429 at commit [`76f4248`](https://github.com/apache/spark/commit/76f424830418129c12a2a08d81f19377490c95eb).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98461/
Test FAILed.

MaxGekk committed:
Merge remote-tracking branch 'origin/master' into plan-to-file

SparkQA commented:
**[Test build #98465 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98465/testReport)** for PR 22429 at commit [`f7de26d`](https://github.com/apache/spark/commit/f7de26d39df135d38b87660e5f68764891106508).

SparkQA commented:
**[Test build #98465 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98465/testReport)** for PR 22429 at commit [`f7de26d`](https://github.com/apache/spark/commit/f7de26d39df135d38b87660e5f68764891106508).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98465/
Test PASSed.

MaxGekk committed:
Merge branch 'master' into plan-to-file

SparkQA commented:
**[Test build #98475 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98475/testReport)** for PR 22429 at commit [`bda6ac2`](https://github.com/apache/spark/commit/bda6ac25882fdff5bf66b845cd6935154f1f744e).

SparkQA commented:
**[Test build #98475 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98475/testReport)** for PR 22429 at commit [`bda6ac2`](https://github.com/apache/spark/commit/bda6ac25882fdff5bf66b845cd6935154f1f744e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98475/
Test PASSed.

MaxGekk commented:
@hvanhovell Could you look at the PR, please.

MaxGekk commented:
@cloud-fan @gatorsmile May I ask you to look at the PR. It stuck for a while by unclear reasons but I believe the proposed method `toFile` could be pretty useful in troubleshooting different issues, and I just don't want to close it so easily.

MaxGekk commented:
@cloud-fan 

> This is hard to review, do you mean we should add maxFields: Option[Int] to all the string related methods?

Not to all but only to methods involved to producing textual representation of different plans. Those are `simpleString` and `verboseString` in children of `TreeNode` + `simpleString` of `StructType`. The changes are trivial - I just made them by fixing compilation errors. Maybe I missed somethings but only because it wasn't touched by compiler. 

Main changes are concentrated in `QueryExecution.scala` where everything begins from the `toFile` method. Other changes are consequences of that. The main idea is to write a OutputStream instead of strings materializations in memory. So, it should allow to avoid limits for string size and OOM which we see sometimes in corner cases.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:
```diff
@@ -176,9 +176,9 @@ case class TakeOrderedAndProjectExec(
-  override def simpleString: String = {
-    val orderByString = Utils.truncatedString(sortOrder, "[", ",", "]")
-    val outputString = Utils.truncatedString(output, "[", ",", "]")
+  override def simpleString(maxFields: Option[Int]): String = {
```
Can we just get rid of the `maxFields`? I think this makes the PR hard to read.

HyukjinKwon commented:
I took a super quick pass - the change actually quite looks okay in general to me.

MaxGekk commented:
@HyukjinKwon @cloud-fan Thank you for looking at the PR.

So, if I split the PR to 2 PRs:
1. Writing truncated plans to a file 
2. Control number of fields in truncated strings.

, it would be better for review, right?

> I think it's okay to expose that number to toFile. I think this is an internal API, right?

Yes, it is.

> People could just set whatever number they want I guess. Fix me if I misread.

I just had a concern that there are much more places (~10) besides of the `toFile` method where I have to pass some value for `maxFields`. In all such places I have to read a SQL config.

HyukjinKwon commented:
@boy-uber, for structured streaming, let's do it out of this PR. I think the actual change of this PR can be small (1.). We can change this API for structured streaming later if needed since this is just an internal private API. Change (2.) can be shared I guess even if we go for structured streaming stuff.

boy-uber commented:
> @boy-uber, for structured streaming, let's do it out of this PR. I think the actual change of this PR can be small (1.). We can change this API for structured streaming later if needed since this is just an internal private API. Change (2.) can be shared I guess even if we go for structured streaming stuff.

Hi @HyukjinKwon thanks for the comment! My original suggestion was not related to structured streaming. It was to support structured logging for Spark plan, for example, get the Spark plan as a json blob, and send it to Spark Listener. This will make it easy to Spark users to write a Spark Listener plugin to get the plan and process the json in their code. Then people could parse and analyze a large number of Spark applications based on their plans. One use case for this is using machine learning to predict resource usage based on the Spark plans. This may not be related to this PR. If you or other people are interested, we could talk offline.

HyukjinKwon commented:
Ooops i rished to read. Yea but still sounds related but orthogonal. Let's move it to mailing list. That should be the best place to discuss further.

MaxGekk commented:
I am closing the PR since a part of it has been merged in #23018 already, and the rest is coming soon.

MaxGekk commented:
Here is a PR introduces `maxFields` parameter to all function involved in creation of truncated strings of spark plans: https://github.com/apache/spark/pull/23159
